design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,350 @@
+and 
+[Kubernetes Authenticator](/app/domain/authentication/authn_k8s/validate_resource_restrictions.rb), 
+there is a lot of code duplication with minor changes. 
+
+## Solution
+
+For the sake of simplicity, from here and throughout the document, Resource
+Restrictions will be defined only in the host's annotations, and not in the ID. 
+In future authenticators we will not develop an option to define it in the ID
+and it is available in the K8s Authenticator only for backwards compatibility.
+
+### Defining Resource Restrictions Validation
+
+The Resource Restrictions validation can be split into 3 parts:
+ 1. Extract the Resource Restrictions from the host
+ 2. Validate the Resource Restrictions are defined correctly
+ 3. Validate the Resource Restrictions match the authentication request
+
+Let's break down each part to better understand them.
+
+#### 1. Extract the Resource Restrictions from the host
+
+This step goes to the host in the policy and extracts the Resource Restrictions
+from its annotations.
+
+Requirements:
+
+* Extract both **granular restrictions** (starts with `authenticator_name`) and
+  **specific restrictions** (starts with `authenticator_name/service_id`).
+* If same restriction exist for both **granual** and **specific** restrictions,
+  **specific** restrictions should prevail.
+* Ignore specific annotations of a different service id
+  (starts with `authenticator_name/other_service_id`).
+* Return a `resource_restrictions` hash of all the restrictions, without their prefix.
+
+In this case the general class will be defined as follows:
+- input
+  - authenticator name
+  - authenticator service-id
+  - account
+  - host name
+- output
+  - `resource_restrictions` hash.
+  
+
+So this class signature is:
+
+```ruby
+ExtractResourceRestrictions = CommandClass.new(
+  dependencies: {
+    role_class:     ::Role,
+    resource_class: ::Resource,
+    logger:         Rails.logger
+  },
+  inputs:   %i(authenticator_name service_id host_name account)
+) do
+
+  def call
+    fetch_all_role_annotations
+    filter_authenticator_annotations
+    convert_to_resource_restrictions
+  end
+
+  private
+  
+  #...
+
+end
+```
+
+
+
+This is the case also in
+[`authn-k8s`](/app/domain/authentication/authn_k8s/resource_restrictions.rb), 
+with only one difference as the K8s Authenticator also looks in the host's ID.
+To support that, this class can be expanded to extract the constraints from the
+host ID, similar to:
+
+```ruby
+ExtractK8sResourceRestrictions = CommandClass.new(
+  dependencies: {
+    role_class:                     ::Role,
+    resource_class:                 ::Resource,
+    extract_resource_restrictions:  ExtractResourceRestrictions.new,
+    logger:                         Rails.logger
+  },
+  inputs:    %i(authenticator_name service_id host_name account)
+) do
+
+  def call
+    extract_resource_restrictions_from_annotations
+    extract_resource_restrictions_from_host if resource_restrictions.empty?
```

Comment:
this should be `extract_resource_restrictions_from_host_id`, correct?

Reply:
done